plugins/s3-viewer-backend/src/credentials-provider/IAMRoleCredentialsProvider.ts

```diff
@@ -0,0 +1,82 @@
+import { LoggerService } from '@backstage/backend-plugin-api';
+import { BucketCredentials, CredentialsProvider, S3Platform } from '../types';
+import { Config } from '@backstage/config';
+import { S3 } from '@aws-sdk/client-s3';
+
+export class IAMRoleCredentialsProvider implements CredentialsProvider {
+  constructor(
+    readonly platforms: S3Platform[],
+    readonly logger: LoggerService,
+    readonly allowedBuckets: { [key: string]: string[] },
+  ) {}
+
+  static fromConfig(
+    config: Config,
+    logger: LoggerService,
+    allowedBuckets: { [key: string]: string[] },
+  ): IAMRoleCredentialsProvider {
+    const platforms: S3Platform[] = config
+      .getConfigArray('platformns')
+      .map(cfg => {
+        const name = cfg.getOptionalString('name') || cfg.getString('endpoint');
+        return {
+          endpoint: cfg.getString('endpoint'),
+          endpointName: name,
+          region: cfg.getString('region'),
+        };
+      });
+
+    return new IAMRoleCredentialsProvider(platforms, logger, allowedBuckets);
+  }
+
+  async getBucketCredentials(): Promise<BucketCredentials[]> {
+    const bucketCreds: BucketCredentials[] = [];
+    await Promise.all(
+      this.platforms.map(async platform => {
+        try {
+          const s3Client = new S3({
+            apiVersion: '2006-03-01',
+            endpoint: platform.endpoint,
+            region: platform.region,
+            forcePathStyle: true,
+          });
+
+          const bucketList = await s3Client.listBuckets({});
+
+          const buckets =
```

Comment:
This code here (and the other filters?) are likely copied over from the other provider, is there a way to share them?

Reply:
I will check a bit more. For config and iam-role most of the code is the same, they could probably share it. I'll check how it can be easily done

Reply:
I've added a second commit with a new file `utils.ts`, which is used by the `config` and `iam-role` providers, since they were sharing most of the code. Let me know what you think